Darwin

--- bsd-asprintf.c.orig	2024-01-19 20:18:17.345301296 +0100
+++ bsd-asprintf.c	2024-01-19 20:18:35.478919520 +0100
@@ -28,13 +28,18 @@
 # undef HAVE_ASPRINTF
 #endif
 
-#ifndef HAVE_VASPRINTF
-
 #include <errno.h>
 #include "bsd-stdarg.h"
 #include <stdio.h>
 #include "bsd-stdlib.h"
 
+#ifdef HAVE_LIMITS_H
+#include <limits.h>
+#endif
+#include <stdio.h>
+
+#ifndef HAVE_VASPRINTF
+
 #define INIT_SZ	128
 
 int
